src/main/scala/firrtl/passes/LowerTypes.scala

```diff
@@ -280,15 +285,17 @@ object LowerTypes extends Transform with DependencyAPIMigration {
       case m: ExtModule =>
         m copy (ports = portsx)
       case m: Module =>
-        m copy (ports = portsx) map lowerTypesStmt(memDataTypeMap, m.info, m.name, renames)
+        m copy (ports = portsx) map lowerTypesStmt(memDataTypeMap, m.info, m.name, renames, initializedMems)
     }
   }
 
   def execute(state: CircuitState): CircuitState = {
+    // TODO: how do we match memories that are not in the top module
+    val initializedMems = state.annotations.collect{case m : MemoryInitAnnotation if !m.isRandomInit => m.target.ref}.toSet
```

Comment:
This needs to be fixed. Maybe someone like @seldridge or @azidar could point me to an example on how to correctly match all annotated memories no matter which module they are in.

Reply:
This does in fact match all annotated memories. However, it is associating them with their declared name only, which is only meaningful in the scope of an individual module. A better approach would be:
```suggestion
    val initializedMems = state.annotations.collect {
      case m : MemoryInitAnnotation if !m.isRandomInit => (m.target.module, m.target.ref)
    }.toSet
```

Reply:
Albert's approach works. I'd probably write this as a `Set[ReferenceTarget]` or `Set[Target]` and then do checking inside each module. This pushes towards a pattern of walking the AST while tracking the target. This assumes, as [Albert mentions](https://github.com/freechipsproject/firrtl/pull/1645#issuecomment-634874207) that the circuit has been properly duplicated such that all memory targets have no instances in their path.

Reply:
@seldridge, an alternative to duplicating before `LowerTypes` is just to consider that "an error in one instance is still an error in the module."

Note the use of encapsulatingModule, which is the actual module containing the reference. The module field of the ReferenceTarget is the root of its cross-module reference path.

I.e, if I wish to target mem m in module t, but only in the instance defined by the hierarchical path q.r.s in module q, the module would be q, while the encapsulatingModule (that actually contains the memory) would be t.